### DISCUSSION TITLE
Elastic moduli as internal variables

### DISCUSSION THREAD

**Discussion by psvay** [thelfer/tfel#626] [Q&A] [answered]
Hello, 

I am a beginner in mfront. I would like to define an (orthotropic) elastic stiffness tensor where the constants are `@AuxiliaryStateVariable`

For context, I need to define a volume element with Amitex, where the elastic constants depend on the position Cij(x). (The most intuitive way to do so could be to define different materials, but since I would like to try different functions Cij(x), it would be way more convenient to declare the elastic moduli as `@AuxiliaryStateVariable`.) 

Is there a way to do it ? I hope my question was clear !

thank you in advance,

Patricia

**Comment by thelfer** [comment]
Hi @psvay,
The usual way to do this is to use material properties. Why would you want to use auxiliary variables ?
The trouble is that most advanced features won't handle elastic properties defined through auxiliary variables.
Thomas

**Reply by psvay** [reply]
Than,k you for your answer. 

I would like to compare different elastic fields for modeling materials subjected to chemical pretreatments.
If the elastic properties can't be defined as auxiliary variables, maybe it could be done indirectly ? I took a look at the damage behaviour files (for example docs/gallery/behaviours/damage/Marigo.mfront), where the stiffness tensor depends on a `@StateVariable` d. 
I was thinking that maybe, for my case, the elastic moduli could be expressed as Cij(d(x)) = d(x)*Cij^0 where d is an `@AuxiliaryVariable`, and Cij^0 are the material properties. Would it be possible ?

**Reply by thelfer** [reply]
The real question is why aren't they defined as material properties ? Is it a limitation of `AMITEX_FFTP` ?
Otherwise, I would recommend passing the coordinates as external state variables.

**Reply by psvay** [reply]
Yes you are right, I don't think there is such limitations so it could be done with amitex ! But the elastic field I want to apply is a gradient, and I would like to try different parameters. So it becomes very difficult and unconvenient to generate volume elements, handling both multiple zones on the composite and the attribution of a high number of materials voxel per voxel. Thank you for your recommandation !

**Reply by psvay** [reply, marked as answer]
Hi @thelfer,

Here is un update on how my problem was solved.
First I would like to clarify why I would want to define elastic properties as internal variables. The reasons for this are that (1) AMITEX can't explicitely define materials with varying mechanical properties, so it would be tedious to do it without Mfront, and (2) in the future, it may be desirable to compute coupling effects making the elastic moduli vary when subject to both diffusion and mechanical loadings (for my case).

As I am using the bricks @OrthotropicBehaviour and @FiniteStrainSingleCrystal, I used @InitLocalVariable before the block @Integrator to modify the elastic moduli, see below. Thank you very much for your answers. I hope I was clear, if not I can provide more details !

```
@AuxiliaryStateVariable real C11;
C11.setEntryName("C11");
@AuxiliaryStateVariable real C33;
C33.setEntryName("C33");
@AuxiliaryStateVariable real C12;
C12.setEntryName("C12");
@AuxiliaryStateVariable real C13;
C13.setEntryName("C13");
@AuxiliaryStateVariable real C44;
C44.setEntryName("C44");
@AuxiliaryStateVariable real C66;
C66.setEntryName("C66");

@InitLocalVariables<Append>{
    D(0,0) = D(0,0)*C11; // C11
    D(1,1) = D(1,1)*C11; // C22
    D(2,2) = D(2,2)*C33; // C33
    D(3,3) = D(3,3)*C44; // C44
    D(4,4) = D(4,4)*C44; // C55
    D(5,5) = D(5,5)*C66; // C66
    D(0,1) = D(0,1)*C12; // C12
    D(1,0) = D(1,0)*C12; // C21
    D(0,2) = D(0,2)*C13; // C13
    D(2,0) = D(2,0)*C13; // C31
    D(1,2) = D(1,2)*C13; // C23
    D(2,1) = D(2,1)*C13; // C32
};
```

**Reply by thelfer** [reply]
@psvay Ok, that's nice.
Did you use the `@RequireStiffnessStensor` keyword ?
